components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+        }
+
+        Ok(outgoing)
+    }
+
+    /// Decrements the change counter, updates the sync status, and cleans up
+    /// tombstones for successfully synced items. Sync calls this method at the
+    /// end of each bookmark sync.
+    fn push_synced_items(&self, uploaded_at: ServerTimestamp, guids: &[String]) -> Result<()> {
+        // Flag all successfully synced records as uploaded. This `UPDATE` fires
+        // the `pushUploadedChanges` trigger, which updates local change
+        // counters and writes the items back to the synced bookmarks table.
+        sql_support::each_chunk(&guids, |chunk, _| -> Result<()> {
+            self.db.execute(
+                &format!(
+                    "UPDATE itemsToUpload SET
+                       uploadedAt = {uploaded_at}
+                     WHERE guid IN ({values})",
+                    uploaded_at = uploaded_at.as_millis(),
+                    values = sql_support::repeat_sql_values(chunk.len())
+                ),
+                chunk,
+            )?;
+            Ok(())
+        })?;
+
+        // Fast-forward the last sync time, so that we don't download the
+        // records we just uploaded on the next sync.
+        put_meta(
+            self.db,
+            LAST_SYNC_META_KEY,
+            &(uploaded_at.as_millis() as i64),
+        )?;
+
+        // Clean up.
+        self.db.execute_batch("DELETE FROM itemsToUpload")?;
+
+        Ok(())
+    }
+}
+
+impl<'a> Store for BookmarksStore<'a> {
+    #[inline]
+    fn collection_name(&self) -> &'static str {
+        "bookmarks"
+    }
+
+    fn apply_incoming(
+        &self,
+        inbound: IncomingChangeset,
+        incoming_telemetry: &mut telemetry::EngineIncoming,
+    ) -> result::Result<OutgoingChangeset, failure::Error> {
+        use dogear::Store;
+
+        // Stage all incoming items.
+        let timestamp = self.stage_incoming(inbound, incoming_telemetry)?;
+
+        // write the timestamp now, so if we are interrupted merging or
+        // creating outgoing changesets we don't need to re-download the same
+        // records.
+        put_meta(self.db, LAST_SYNC_META_KEY, &(timestamp.as_millis() as i64))?;
+
+        // Merge and stage outgoing items.
+        self.merge_with_driver(&Driver)?;
+
+        let outgoing = self.fetch_outgoing_records(timestamp)?;
+        Ok(outgoing)
+    }
+
+    fn sync_finished(
+        &self,
+        new_timestamp: ServerTimestamp,
+        records_synced: &[String],
+    ) -> result::Result<(), failure::Error> {
+        let tx = self.db.unchecked_transaction()?;
+        let result = self.push_synced_items(new_timestamp, records_synced);
+        match result {
+            Ok(_) => tx.commit()?,
+            Err(_) => tx.rollback()?,
+        }
+        result?;
+        Ok(())
+    }
+
+    fn get_collection_request(&self) -> result::Result<CollectionRequest, failure::Error> {
+        let since = get_meta::<i64>(self.db, LAST_SYNC_META_KEY)?
```

Comment:
Desktop's logic for this is [a bit fancier](https://searchfox.org/mozilla-central/rev/f1c7ba91fad60bfea184006f3728dd6ac48c8e56/toolkit/components/places/SyncedBookmarksMirror.jsm#241-265), because it tries to handle the case where the sync was interrupted as we were downloading new items. It's only likely to come up for first syncs, or for clients that haven't synced in a while. For interrupted large batches, all records will have the same timestamp, so we'll still download them again.